posthog/hogql_queries/insights/funnels/funnel_udf.py

```diff
@@ -0,0 +1,158 @@
+
+
+class FunnelUDF(FunnelBase):
+    def get_step_counts_query(self):
+        max_steps = self.context.max_steps
+        return self._get_step_counts_query(
+            outer_select=[
+                *self._get_matching_event_arrays(max_steps),
+            ],
+            inner_select=[
+                *self._get_matching_events(max_steps),
+            ],
+        )
+
+    def conversion_window_limit(self) -> int:
+        return int(
+            self.context.funnelWindowInterval * DATERANGE_MAP[self.context.funnelWindowIntervalUnit].total_seconds()
+        )
+
+    def get_query(self) -> ast.SelectQuery:
+        if self.context.funnelsFilter.funnelOrderType == "strict":
+            inner_event_query = self._get_inner_event_query_for_udf(
+                entity_name="events", skip_step_filter=True, skip_entity_filter=True
+            )
+        else:
+            inner_event_query = self._get_inner_event_query_for_udf(entity_name="events")
+
+        default_breakdown_selector = "[]" if self._query_has_array_breakdown() else "''"
+
+        # stores the steps as an array of integers from 1 to max_steps
+        # so if the event could be step_0, step_1 or step_4, it looks like [1,2,0,0,5]
+
+        # Each event is going to be a set of steps or it's going to be a set of exclusions. It can't be both.
+        steps = ",".join([f"{i + 1} * step_{i}" for i in range(self.context.max_steps)])
+
+        # this will error if they put in a bad exclusion
+        exclusions = ""
+        if getattr(self.context.funnelsFilter, "exclusions", None):
+            exclusions = "".join([f",-{i + 1} * exclusion_{i}" for i in range(1, self.context.max_steps)])
+
+        if self.context.breakdownType == BreakdownType.COHORT:
+            fn = "aggregate_funnel_cohort"
+            breakdown_prop = ", prop"
+        elif self._query_has_array_breakdown():
+            fn = "aggregate_funnel_array"
+            breakdown_prop = ""
+        else:
+            fn = "aggregate_funnel"
+            breakdown_prop = ""
+
+        prop_selector = "prop" if self.context.breakdown else default_breakdown_selector
+        prop_vals = "groupUniqArray(prop)" if self.context.breakdown else f"[{default_breakdown_selector}]"
+
+        breakdown_attribution_string = f"{self.context.breakdownAttributionType}{f'_{self.context.funnelsFilter.breakdownAttributionValue}' if self.context.breakdownAttributionType == BreakdownAttributionType.STEP else ''}"
+
+        inner_select = parse_select(
+            f"""
+            SELECT
+                arrayJoin({fn}(
+                    {self.context.max_steps},
+                    {self.conversion_window_limit()},
+                    '{breakdown_attribution_string}',
+                    '{self.context.funnelsFilter.funnelOrderType}',
+                    {prop_vals},
+                    arraySort(t -> t.1, groupArray(tuple(toFloat(timestamp), {prop_selector}, arrayFilter((x) -> x != 0, [{steps}{exclusions}]))))
```

Comment:
Optimization idea: Could we turn this UDF into a stateful (aggregate) function, so that we don't have to allocate this array?